components/HTTP/src/Websocket.cpp

```diff
@@ -0,0 +1,303 @@
+            ESP_LOGE(TAG, "Error sending message %d", err);
+            return;
+        };
+
+        break;
+    default:
+        break;
+    }
+}
+
+int Websocket::getTypeMessage(uint8_t* payload)
+{
+    char* res = reinterpret_cast<char*>(payload);
+    cJSON* req_root = cJSON_Parse(res);
+    cJSON* type = cJSON_GetObjectItemCaseSensitive(req_root, "type");
+    if (cJSON_IsNumber(type)) {
+        return type->valueint;
+    } else {
+        ESP_LOGE(TAG, "type must be numeric");
+        return -1;
+    }
+}
+
+
+esp_err_t Websocket::getClientData(uint8_t* payload, client_data_t* client)
+{
+    char* res = reinterpret_cast<char*>(payload);
+    cJSON* req_root = cJSON_Parse(res);
+    cJSON* cjson_timestamp = cJSON_GetObjectItemCaseSensitive(req_root, "timestamp");
+    cJSON* cjson_shaUID = cJSON_GetObjectItemCaseSensitive(req_root, "shaUID");
+
+    const char* sha_uid = cjson_shaUID->valuestring;
+    int timestamp = cjson_timestamp->valueint;
+    if (cJSON_IsNumber(cjson_timestamp) && cJSON_IsString(cjson_shaUID)) {
+        if (strlen(sha_uid) != 64) {
+            return ESP_FAIL;
+        }
+
+        util::hexToBytes(sha_uid, client->shaUID);
+
+        client->timestamp = timestamp;
+        client->is_alive = true;
+        client->fd = httpd_req_to_sockfd(req_handler);
+        return ESP_OK;
+    };
+
+    return ESP_FAIL;
+}
+
+
+void ws_async_send(void* arg)
+{
+    static const char* data = "Async data";
+    struct async_resp_arg_t* resp_arg = (struct async_resp_arg_t*)arg;
+    httpd_handle_t hd = resp_arg->hd;
+    int fd = resp_arg->fd;
+    httpd_ws_frame_t ws_pkt;
+    memset(&ws_pkt, 0, sizeof(httpd_ws_frame_t));
+    ws_pkt.payload = (uint8_t*)data;
+    ws_pkt.len = strlen(data);
+    ws_pkt.type = HTTPD_WS_TYPE_TEXT;
+
+    std::cout << ws_pkt.payload << std::endl;
+
+    httpd_ws_send_frame_async(hd, fd, &ws_pkt);
+    free(resp_arg);
+}
+
+esp_err_t Websocket::trigger_async_send(httpd_handle_t handle, int fd)
+{
+    struct async_resp_arg_t* resp_arg = (struct async_resp_arg_t*)malloc(sizeof(struct async_resp_arg_t));
+    resp_arg->hd = handle;
+    resp_arg->fd = fd;
+    return httpd_queue_work(handle, ws_async_send, resp_arg);
+}
+
+esp_err_t Websocket::messageRecipient(uint8_t* payload, uid_message_t* uid_receiving, const char* null_to_uid)
+{
+    char* res = reinterpret_cast<char*>(payload);
+    cJSON* req_root = cJSON_Parse(res);
+    cJSON* cjson_fromUID = cJSON_GetObjectItemCaseSensitive(req_root, "fromUID");
+    cJSON* cjson_toUID = cJSON_GetObjectItemCaseSensitive(req_root, "toUID");
+
+    const char* from_uid = cjson_fromUID->valuestring;
+    const char* to_uid = cjson_toUID->valuestring;
+
+    if (cJSON_IsNull(cjson_toUID)) {
+        util::hexToBytes(null_to_uid, uid_receiving->to_uid);
+
+    } else if (cJSON_IsString(cjson_toUID)) {
+        if (strlen(to_uid) != 64) {
+            ESP_LOGI(TAG, "toUID length  is less than 64 ");
+            return ESP_FAIL;
```

Comment:
optionally you can use `ESP_ERR_INVALID_SIZE` in this case is more accurate:
```suggestion
            return ESP_ERR_INVALID_SIZE;
```